### ISSUE TITLE
Server-side search results

### ISSUE THREAD

**Issue by johanbaaij** [mrichar1/jsonapi-vuex#55]
I'm feeding a data-table straight from the store with the `get` getter. Am I correct that it's not possible to entirely replace a type with an API response? Could not find anything in the source so I wrote my own mutation for this but would like to clear the previous results once the promise gets resolved, not before the API is called. This could be useful for server-side search results.

Another option I'm considering is to just store the search results separately and use these when in search mode.

Curious how other people are approaching this.

**Comment by mrichar1** [comment]
If I'm reading this right, you'd like to be able to do e.g. `GET /widgets` and the resulting store contains *only* the 'widgets' items returned from the API response, and any in the store, but not in the response, are dropped. And/or be able to query the API without actually changing the state of the store.

Currently there aren't mutations/actions to handle these mode, but I'm not averse to adding this functionality if it is useful!

**Comment by johanbaaij** [comment]
Correct, or something like `GET /articles?filter[text_contains]=string`. 

Querying the API without changing the store isn't a necessity for my case. All I want to do is filter a datatable. My mutation is really simple, it just replaces everything in the `article` node with an empty object (`state[type] = {}`).

What I would like is this to happen roughly before https://github.com/mrichar1/jsonapi-vuex/blob/4002f76e76268c4fb1ed9ff12c2d0d532afbf21c/src/jsonapi-vuex.js#L82

Maybe I'm overlooking things and it's just not that simple :). For example, maybe any related objects should be removed from the store.

Server-side sorting might be a bit of a challenge too? The order of the `data` array that an API is returning isn't preserved anywhere. To complicate things, I fetch subsequent pages by using the `links.next` property of the API response.

**Comment by staaam** [comment]
Is there a necessity to go through the store? As an option you could fetch results straight from the response:
```javascript
context.dispatch('jv/get', rss)
  .then(result  => { <store search result in "private" store"> })
```

**Comment by mrichar1** [comment]
Ah thanks - that makes things a bit clearer.

I think this can be broken down into 3 separate issues/features (while you personally don't need all 3 it's worth considering them all anyway as they all have their merits...)

1. Actions that return data but don't change the store.
2. Store exactly mirrors the latest response (per-type).
3. Sorting of data.

1. Is relatively easy to implement, since it is just `get` without the `commit` parts. I'll look into how to do this - I have a feeling that trying to make `get` take this as a parameter would be messy, and it would be better to implement a new action - maybe called `search` or similar?

2. This is also relatively easy - I think the correct way to implement this (if enabled) is to first update the store from the response data, then afterwards delete existing records from the store that weren't in the response. That way the store is never completely empty and we don't delete and then add back identical records (both of which would cause properties to be re-computed and potentially result in UI redraws).

3. Sorting is trickier. It is apparently the case that objects now have a defined property order, which might mean the store could be used as-is (provided care is taken to ensure they are populated in order from the API response): https://kangax.github.io/compat-table/es6/#test-own_property_order  If that isn't viable then the next option would be to store some meta-data about the last request (under the top-level `_jv` which contains references to the resources in order, per-type). This could then be used directly, or via computed properties/helper methods.

-Let me know what you think about the above - I'll have a play around with 3. and see if jsonapi sorting  is reproduceable just using object ordering before committing to any work in that area either way.

**Comment by johanbaaij** [comment]
That sounds good me. Your implementation of 2. seems a lot nicer too.

I've been trying to work out what a pattern for the combination of sorting and pagination could look like. It would get too messy to aggregate sorting of multiple pages, there could be gaps if pages are skipped, user could change the sorting direction, etc. 

Maybe I'm trying to hold on too much to the idea that if any data has already been received from the API it should be prevented to make an extra call. In this case it's just so much easier to simply call the API again and let browser/server caching do the lifting. 

Hope those last two paragraphs make sense. Let me know if I can be of more help.

**Comment by mrichar1** [comment]
#57 adds support for 2. - namely to ensure that each `type` in the store exactly matches the records received in the last response. Let me know if this is what you were looking for.

I'll start work on 1. next...

**Comment by mrichar1** [comment]
#58 adds support for 1. i.e it calls the `get` action but with `context.commit` replaced with an empty function, so that no store mutations can be called.

**Comment by mrichar1** [comment]
Fixed by #57 and #58.
